### ISSUE TITLE
Rendering Queue issue

### ISSUE THREAD

**Issue by neohun** [ColinLeung-NiloCat/UnityURPUnlitScreenSpaceDecalShader#6]
Hello first of all thank you for sharing this decal shader it's amazing and you can't know how much time you have saved for me otherwise I had to create my own shader via shader graph in unity and this would take a while.. When I release my game I'll credit you properly for your work.. So here's the issue I wanna mention: the decal materials which use this shader is rendered properly If I set the "Render Queue" 2501 and above (as default it's 2451 and I have changed it in tags) if the value is less than this value decal doesn't show up on the "Game View" and I have to set my characters higher than this to don't allow the decals override my characters so basically I want this shader to work between 2451 - 2499 somewhere do you have any idea about how to fix that And also I should mention that if you enable "msaa antialiasing" from the "URP Asset" settings it works with "2451" render queue value.. Here's two screen-shot explains what I mean.. I'll be waiting for your reply, 
Best Regards, 
![ss_2501](https://user-images.githubusercontent.com/63730013/79352793-f698ce80-7f42-11ea-8c61-8751f0bc5a67.PNG)
![ss_2451](https://user-images.githubusercontent.com/63730013/79352813-f993bf00-7f42-11ea-8936-da6c0f7a9801.PNG)

**Comment by ColinLeung-NiloCat** [comment]
so if MSAA on
= queue 2451 & 2501 works fine? Don't have any problem?
----------------------------
but if MSAA off
= queue 2451 doesn't work?
= queue 2501 works?

**Comment by neohun** [comment]
> ## so if MSAA on
> = queue 2451 & 2501 works fine? Don't have any problem?
> but if MSAA off
> = queue 2451 doesn't work?
> = queue 2501 works?

Yeah Exactly.. That's why it works with URP template project default settings because the msaa is on as default..

**Comment by neohun** [comment]
And when I have made a bug report because I was thinking that it's a unity bug then they have sent me this answer I'll share here if that makes sense for you I'm not sure if this issue is related with this or not..

"After further investigating your issue we have found that the problem is in your custom shaders. When using clip(mask-0.5), mask is a bool, when clip() is expecting a float. The reason this worked before is that the compiler "Accidentaly" converted the values to the right types. In the newer version, some changes were made, that made this accidental behavior non-existent.

Inspecting the code down to assembly such an issue is then found:

When lt instruction is ran, it returns true(0xFFFFFFFF) or a false(0x0000000), but because it gets a bool(it is expecting a float) it returns NaN. Finally it comes down to discard instruction and when it is expecting a true or a false, it gets NaN which means that the shaders will not be rendered correctly. "

clip():
https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-clip

lt instruction:
https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/lt--sm4---asm-

discard instruction:
https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/discard--sm4---asm-

**Comment by ColinLeung-NiloCat** [comment]
I tried a fix

**Comment by neohun** [comment]
unfortunately it didn't make any difference so somehow it works when it enters the transparent queue range.. (above 2501)

**Comment by ColinLeung-NiloCat** [comment]
really no idea, I don't even know what to google, maybe you should find a more professional shader developer here
https://forum.unity.com/threads/released-lux-lwrp-essentials.712619/
his asset has a decal shader also.

**Comment by neohun** [comment]
Alright thanks for your interest If I find a viable solution I'll share here..

**Comment by neohun** [comment]
Alright I have fixed the problem but for make this shader to work in every condition it should be rendered in transparent render queue (great equal 2501) so I think you should change it from the tags for an example for  "Queue"="Transparent-250" can be good because I have also checked the other lit and unlit decals they set it to transparent queue so it's not a bug it's just which order the unity renders the objects so just set as transparent to work properly..
And for those who wanna learn how I fixed the decal override problem here's the linked below for detailed info..

https://forum.unity.com/threads/solved-urp-depth-of-field-transparency-render-queue-problem.859936/

**Comment by ColinLeung-NiloCat** [comment]
You are correct, I checked the doc and the doc wrote this
"Queues up to 2500 (“Geometry+500”) are considered “opaque” and optimize the drawing order of the objects for best performance. Higher rendering queues are considered for “transparent objects” and sort objects by distance, starting rendering from the furthest ones and ending with the closest ones. Skyboxes are drawn in between all opaque and all transparent objects."

So, when queue is <2501, it is still not safe to draw decal yet, because the depth buffer can still be changed.
I now changed the default queue to 2501, which all depth write must be finished = safe to draw decal.

Thanks for your help!
